Don't call `Rails.start()` in automatically when `@rails/ujs` is imported as ESM

tricknotes opened the pull request:
### Motivation / Background

<!--
Describe why this Pull Request needs to be merged. What bug have you fixed? What feature have you added? Why is it important?
If you are fixing a specific issue, include "Fixes #ISSUE" (replace with the issue number, remove the quotes) and the issue will be linked to this PR.
-->

This Pull Request has been created because https://github.com/rails/rails/issues/49499 should be fixed in Rails 7.1.2.

### Detail

This Pull Request changes auto start logic for `@rails/ujs` as ESM.
This start logic has been changed since https://github.com/rails/rails/pull/45546 unexpectedly.

### Additional information

In the spec of ECMAScript, `this` should be `undefined`.

> 8.1.1.5 Module Environment Records
> 8.1.1.5.4 `GetThisBinding` ()
> 1. Return `undefined`.

Ref: https://262.ecma-international.org/8.0/#sec-module-environment-records-getthisbinding

Since some of build tools (e.g. esbuild) don't have `exports` nor `module` object, adding `this` checking is required whether this file is imported as esm or not.

Close https://github.com/rails/rails/issues/49499

### Checklist

Before submitting the PR make sure the following are checked:

* [x] This Pull Request is related to one change. Changes that are unrelated should be opened in separate PRs.
* [x] Commit message has a detailed description of what changed and why. If this PR fixes a related issue include it in the commit message. Ex: `[Fix #issue-number]`
* [ ] Tests are added or updated if you fix a bug or add a feature.
* [x] CHANGELOG files are updated for the changed libraries if there is a behavior change or additional feature. Minor bug fixes and documentation changes should not be included.

Earlopain commented:
This doesn't seem to work quite right. Setting a breakpoint at the appropriate place and checking `typeof this` returns `object`.
![image](https://github.com/rails/rails/assets/14981592/6fb83c29-0899-49f2-bcbb-66b890605e62)

tricknotes commented:
Thanks for your confirmation, @Earlopain .
Yes, your debug is the same as my expectation.
When `this` is not `undefined`, the `Rails` object should be exported into `window` object.

Could you tell me more about the problem cases?

tricknotes commented:
Until I fixed tests, I keep this PR as draft 🙏

tricknotes committed:
Don't call `Rails.start()` in automatically when `@rails/ujs` is imported as esm.

In the spec of ECMAScript, `this` should be  `undefined`.

> 8.1.1.5 Module Environment Records
> 8.1.1.5.4 `GetThisBinding` ( )
> 1. Return `undefined`.

Ref: https://262.ecma-international.org/8.0/#sec-module-environment-records-getthisbinding

Since some of build tools (e.g. esbuild) don't have `exports` nor `module`
object, adding `this` checking is required whether this file is imported as esm or not.

Close https://github.com/rails/rails/issues/49499

Earlopain commented:
> Thanks for your confirmation, @Earlopain . Yes, your debug is the same as my expectation. When `this` is not `undefined`, the `Rails` object should be exported into `window` object.
> 
> Could you tell me more about the problem cases?

I did a checkout of this code and I realize that my mental model of how this all works is wrong. Esbuild has transpiled the new check down to just `false` so the point I made earlier about the window object is moot since it doesn't even run in that context. Please ignore my comment.
